libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1719,12 +1823,79 @@ std::optional<std::pair<PeriodData, std::vector<std::shared_ptr<Vote>>>> PbftMan
     }
   }
 
-  return std::optional<std::pair<PeriodData, std::vector<std::shared_ptr<Vote>>>>(
+  const auto kBlockPeriod = period_data.pbft_blk->getPeriod();
+  const auto extra_data = period_data.pbft_blk->getExtraData();
+  if (kGenesisConfig.state.hardforks.ficus_hf.isFicusHardfork(kBlockPeriod)) {
+    if (!extra_data.has_value()) {
+      LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << kBlockPeriod
+                   << " does not contain extra data";
+      sync_queue_.clear();
+      net->handleMaliciousSyncPeer(node_id);
+      return std::nullopt;
+    }
+
+    // Validate pillar block hash
+    if (kGenesisConfig.state.hardforks.ficus_hf.isPbftWithPillarBlockPeriod(kBlockPeriod)) {
+      if (!extra_data->getPillarBlockHash().has_value()) {
+        LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << kBlockPeriod
+                     << " does not contain pillar block hash";
+        sync_queue_.clear();
+        net->handleMaliciousSyncPeer(node_id);
+        return std::nullopt;
+      }
+    } else {
+      if (extra_data->getPillarBlockHash().has_value()) {
+        LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << period_data.pbft_blk->getPeriod()
+                     << " contains pillar block hash";
+        sync_queue_.clear();
+        net->handleMaliciousSyncPeer(node_id);
+        return std::nullopt;
+      }
+    }
+
+    // Validate pillar votes
+    if (kGenesisConfig.state.hardforks.ficus_hf.isPillarBlockPeriod(kBlockPeriod, 2)) {
+      if (!period_data.pillar_votes_.has_value()) {
+        LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << kBlockPeriod
+                     << " does not contain pillar votes";
+        sync_queue_.clear();
+        net->handleMaliciousSyncPeer(node_id);
+        return std::nullopt;
+      }
+
+      // Validate pillar votes
+      if (!validatePbftBlockPillarVotes(period_data)) {
+        LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << kBlockPeriod
+                     << " doesn't have enough valid pillar votes. Clear synced PBFT blocks!";
+        sync_queue_.clear();
+        net->handleMaliciousSyncPeer(node_id);
+        return std::nullopt;
+      }
+    } else {
+      if (period_data.pillar_votes_.has_value()) {
+        LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << period_data.pbft_blk->getPeriod()
+                     << " contains pillar votes";
+        sync_queue_.clear();
+        net->handleMaliciousSyncPeer(node_id);
+        return std::nullopt;
+      }
+    }
+  } else {
+    if (extra_data.has_value()) {
+      LOG(log_er_) << "Synced PBFT block " << pbft_block_hash << ", period " << kBlockPeriod
+                   << " contains extra data even though it should not";
+      sync_queue_.clear();
+      net->handleMaliciousSyncPeer(node_id);
+      return std::nullopt;
+    }
```

Comment:
this field is used for whatever extra data we want, including pillar block hash

Reply:
hmm not sure ? extra_data was implemented by Miho or Dmitry I think and we use it since ficus_hf. There are no checks when constructing pbft block from rlp so even if someone includes whatever in extra data before ficus_hf, it will work... I am then making sure in block validation that extra_data is not used before ficus_hf